### PR TITLE
added signal when sync complete

### PR DESCRIPTION
**Pull request by jodihum**
This looks like a lot of changes but it it because I did pod update.  The only files in the SDK that I changed were:

ConversationClient.swift
NXMSyncManager.swift

Also changed the Podfile and scheme, and changed the project file to use swift 3.0 instead of 3.0.1

### CODE REVIEW

File: NexmoConversation/Core/NXMSyncManager.swift [hunk 1]
```diff
@@ -30,6 +30,7 @@ internal class NXMSyncManager {
      An event to indicate our connected status.
      */
     public let syncConnectionStatusChanged = NXMSignal<(/* status */ NXMConnectionStatus)>()
+    public let syncComplete = NXMSignal<NXMConversationCollection>()
```
Comment (shams-ahmed) [inline NexmoConversation/Core/NXMSyncManager.swift]: 1. `NXMSyncManager` class is internal. can this be changed to internal along with `syncConnectionStatusChanged`

2. We should avoid emitting conversation collection and force the developer to use the collection in client class, I was thinking of a future proof solution such as state enum (started, complete, paused, error) but i think the codebase needs a lot of work before we can do this. If we need something asap could we do: `NXMSignal<Void>()`

File: NexmoConversation/Core/NXMSyncManager.swift [hunk 2]
```diff
@@ -30,6 +30,7 @@ internal class NXMSyncManager {
      An event to indicate our connected status.
      */
     public let syncConnectionStatusChanged = NXMSignal<(/* status */ NXMConnectionStatus)>()
```
Comment (shams-ahmed) [inline NexmoConversation/Core/NXMSyncManager.swift]: change to internal and clean up above. see below comment for reason

File: NexmoConversation/Core/NXMSyncManager.swift [hunk 3]
```diff
@@ -113,6 +114,7 @@ internal class NXMSyncManager {
                     
                     try syncConversation(uuid: uuid)
                 }
+                syncComplete.emit((ConversationClient.instance.conversations))
```
Comment (shams-ahmed) [inline NexmoConversation/Core/NXMSyncManager.swift]: if using the void approach:  `syncComplete.emit()`

File: NexmoConversation/APIObjects/ConversationClient.swift [hunk 4]
```diff
@@ -526,6 +558,10 @@ public class ConversationClient: NSObject {
     private func handleReceivedNewEvent(_ eventType: String, data: JSON) {
         syncManager.handleWebsocketEventReceived(eventType: eventType, data: data)
     }
+    
+    private func handleSyncComplete(collection:NXMConversationCollection) {
```
Comment (shams-ahmed) [inline NexmoConversation/APIObjects/ConversationClient.swift]: No need for us to handle on SDK side, don't think it efficient to emit two times. i would prefer to remove the whole method for a lazy approach. see above comment

File: NexmoConversation/APIObjects/ConversationClient.swift [hunk 5]
```diff
@@ -251,6 +282,7 @@ public class ConversationClient: NSObject {
         // TOOD: research using a global observation where other objects can listen to changes i.e applicationState in AppLifeCycle
         /* Add handlers to the Sync Manager we've just created. */
         syncManager.syncConnectionStatusChanged.addHandler(self, handler: ConversationClient.handleSyncConnectionStatusChanged)
+        syncManager.syncComplete.addHandler(self, handler: ConversationClient.handleSyncComplete)
```
Comment (shams-ahmed) [inline NexmoConversation/APIObjects/ConversationClient.swift]: remove this if taking the lazy property approach.

File: NexmoConversation/APIObjects/ConversationClient.swift [hunk 6]
```diff
@@ -162,6 +162,37 @@ public class ConversationClient: NSObject {
...
+     
+     ```
+     func handleSyncComplete(conversations: NXMConversationCollection) {
+     
+     // Do something here
+     
+     }
+     ```
+     */
+     public let syncComplete = NXMSignal<NXMConversationCollection>()
```
Comment (shams-ahmed) [inline NexmoConversation/APIObjects/ConversationClient.swift]: use the existing signal with: 
```
public lazy var syncComplete: NXMSignal<Void> = { [unowned self] in
        return self.syncManager.syncComplete
}()
```

Note: `syncManager` property needs to be changed to `let`. actually on the todo list
in login method assign and close sync needs to be removed

**Comment by shams-ahmed** [comment]
having looked at the sync manager, that class really needs to be rewritten from scratch. Bad design which kind of led to us doing a signal for sync complete. i doubt it worth the effect at this point in time.

File: NexmoConversation/Core/NXMSyncManager.swift (hunk 2 above)
Comment (jodihum) [inline NexmoConversation/Core/NXMSyncManager.swift]: makes sense

File: NexmoConversation/Core/NXMSyncManager.swift (hunk 1 above)
Comment (jodihum) [inline NexmoConversation/Core/NXMSyncManager.swift]: Also makes sense

File: NexmoConversation/Core/NXMSyncManager.swift (hunk 3 above)
Comment (jodihum) [inline NexmoConversation/Core/NXMSyncManager.swift]: yup

File: NexmoConversation/APIObjects/ConversationClient.swift (hunk 6 above)
Comment (jodihum) [inline NexmoConversation/APIObjects/ConversationClient.swift]: OK with the lazy property.  I am going to leave the syncManager assign and sync for now until we have more time to do it carefully and test it out properly

File: NexmoConversation/APIObjects/ConversationClient.swift (hunk 5 above)
Comment (jodihum) [inline NexmoConversation/APIObjects/ConversationClient.swift]: yup

File: NexmoConversation/APIObjects/ConversationClient.swift (hunk 4 above)
Comment (jodihum) [inline NexmoConversation/APIObjects/ConversationClient.swift]: Yes I also wasn't crazy about emitting twice.  Yours is a better solution.  Thanks.

**Comment by jodihum** [comment]
OK I have made the changes you recommended.  Please take another look.  Thanks.

**Comment by shams-ahmed** [comment]
One last thing, in conversation client class:

- 18: change sync manager to let property. main reason behind this is sync manager gets recreated in the login callback. if a developer observes the syncComplete before logging in then it will become nil soon after. now that i've written this, it explain one issue Sid is having where he gets zero conversation. 
 
- 274-278: remove line since we made it a let

**Comment by jodihum** [comment]
- Made changes requested in PR
- Updated ConversationLite to work properly with new merged version of CAPI
- Changed ourMemberRecord in NXMConversation to prefer a JOINED member
- Added ourMemberRecords to NXMConversation in case all members are required instead of only one
- Modified previous usages of ourMemberRecord accordingly (in NXMSyncManager, NetworkController, and NXMConversation)
- fixed EventState to match with JSON format received and added extension to Gloss's Decoder to parse json in format received [String:Date]  (this solves the problem we were having with too many delivered/seen receipts being sent)

File: NexmoConversation/Core/REST/NetworkController.swift [hunk 7]
```diff
@@ -271,7 +271,7 @@ internal class NetworkController {
     internal func IndicateStatus(message: NXMTextEvent, type: NXMEventBase.NXMRawEventType, callback: @escaping NetworkControllerIndicateStatusCallback) {
         let parameters: [String:Any] = [
             "cid": message.conversation.uuid,
-            "from": message.conversation.ourMemberRecord.uuid,
+            "from": message.conversation.ourMemberRecord?.uuid ?? "noMemberId",
```
Comment (shams-ahmed) [inline NexmoConversation/Core/REST/NetworkController.swift]: Question: what happen when there no noMemberId, does the backend still accept it?

Comment (jodihum) [inline NexmoConversation/Core/REST/NetworkController.swift]: That is a very good question. Would be best just not to send the request in that case.  I will change it.

**Pull request merged**